kafka/mockcluster.go

```diff
@@ -0,0 +1,73 @@
+package kafka
+
+import "C"
+import "unsafe"
+
+/*
+#include <stdlib.h>
+#include "select_rdkafka.h"
+#include "glue_rdkafka.h"
+*/
+import "C"
+
+type MockCluster struct {
+	handle   handle
+	mcluster *C.struct_rd_kafka_mock_cluster_s
+}
+
+/*NewMockCluster provides a mock Kafka cluster with a configurable
+ *number of brokers that support a reasonable subset of Kafka protocol
+ *operations, error injection, etc.
+ *
+ *Mock clusters provide localhost listeners that can be used as the bootstrap
+ *servers by multiple rd_kafka_t instances.
+ *
+ * Currently supported functionality:
+ *  - Producer
+ *  - Idempotent Producer
+ *  - Transactional Producer
+ *  - Low-level consumer
+ *  - High-level balanced consumer groups with offset commits
+ *  - Topic Metadata and auto creation
+ *
+ * @remark High-level consumers making use of the balanced consumer groups
+ *         are not supported.
+ *
+ * @remark This is an experimental public API that is NOT covered by the
+ *         librdkafka API or ABI stability guarantees.
+ *
+ *
+ * @warning THIS IS AN EXPERIMENTAL API, SUBJECT TO CHANGE OR REMOVAL.
+ */
+func NewMockCluster(count int) (*MockCluster, error) {
+
+	mc := &MockCluster{}
+
+	cErrstr := (*C.char)(C.malloc(C.size_t(256)))
+	defer C.free(unsafe.Pointer(cErrstr))
+
+	cConf := C.rd_kafka_conf_new()
+
+	mc.handle.rk = C.rd_kafka_new(C.RD_KAFKA_PRODUCER, cConf, cErrstr, 256)
+	if mc.handle.rk == nil {
+		return nil, newErrorFromCString(C.RD_KAFKA_RESP_ERR__INVALID_ARG, cErrstr)
```

Comment:
cConf is leaked here, do:
`C.rd_kafka_conf_destroy(cConf)`

Reply:
This is now within the close func

Reply:
rd_kafka_new() is a bit weird in that it eats the cConf on success, but not on failure.
So you only need to destroy() on failure.